Update get_it_impl.dart to dispose in reverse order

bvoq opened the pull request:
There are a few situations, where one instance depends on an instance you created before.
In all of these situations, it makes more sense to dispose of the widgets in reverse order.

escamoteur commented on lib/get_it_impl.dart:
```diff
@@ -309,7 +309,7 @@ class _Scope {
   Future<void> reset({required bool dispose}) async {
     if (dispose) {
-      for (final factory in allFactories) {
+      for (final factory in allFactories.reversed) {
```
Hi, thanks for the PR.
I fear that is not enough, as 

```dart
  List<_ServiceFactory> get allFactories => factoriesByName.values
      .fold<List<_ServiceFactory>>([], (sum, x) => sum..addAll(x.values));
``` 

which adds values of maps, which give no guarantee that they are ordered in any meaningful way.
if we want to add a function like this, we will have to add a separate list in every scope where we add a reference to all _ServiceFactory that is added. 
and we also need to add some test that verifies that we do indeed dispose in the reverse order.

bvoq replied:
Perhaps an easier way would be to use LinkedHashMap instead, which is order-preserving by default.
Would you be interested in such a PR?

escamoteur commented:
Hey, after reading the docs of the linked hash map it seems that that would solve the problem without hurting performance.
Good idea
Cheers
Thomas
Am 22. Nov. 2023, 10:00 +0100 schrieb Kevin De Keyser ***@***.***>:
> @bvoq commented on this pull request.
> In lib/get_it_impl.dart:
> > @@ -309,7 +309,7 @@ class _Scope {
>
>   Future<void> reset({required bool dispose}) async {
>     if (dispose) {
> -      for (final factory in allFactories) {
> +      for (final factory in allFactories.reversed) {
> Perhaps an easier way would be to use LinkedHashMap instead, which is order-preserving by default.
> Would you be interested in such a PR?
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>

bvoq commented:
Alright I've added tests. I wouldn't rename the type of `factoriesByName`, since:
`final factoriesByName = <String?, Map<Type, _ServiceFactory<Object, dynamic, dynamic>>>{};` is of type:

`Map<String?, Map<Type, _ServiceFactory<Object, dynamic, dynamic>>>`

which is not of subtype:
```
// ignore: prefer_collection_literals
final factoriesByName = LinkedHashMap<String?,LinkedHashMap<Type, _ServiceFactory<Object, dynamic, dynamic>>>();
```

In this way, this could be released as a minor update.
If we rename the type it might break the clients and would require a major version update.

I'm ok with it either way, do you want to add the CHANGELOG.md?

Thanks for looking at the PR!

escamoteur commented:
Why do you think a change of the type of this  this internal data structure might break any apps that use get_it? Change log addition would be fine. Also do we already have a note in the docs pointing out that we now dispose in reverse order?

bvoq commented:
> Why do you think a change of the type of this this internal data structure might break any apps that use get_it?
I was wrong, my bad. I didn't see that it was part of a private class _Scope.
I've changed the type now and made the changelog more explicit.
I've also added information to the docs pointing out that they are disposed in the reverse order.

escamoteur commented:
thanks a lot for this fine PR, I just published a new version
